### PR TITLE
[MLIR][SCF] Define `-scf-rotate-while` pass

### PR DESCRIPTION
**Pull request by victor-eds**
Define SCF dialect patterns rotating `scf.while` loops leveraging existing `mlir::scf::wrapWhileLoopInZeroTripCheck`. `forceCreateCheck` is always `false` as the pattern would lead to an infinite recursion otherwise.

This pattern rotates SCF.WhileOp, mutating them from "while" loops to "do-while" loops. A guard checking the condition for the first iteration is inserted. Note this guard can be optimized away if the compiler can prove the loop will be executed at least once.

Using this pattern, the following while loop:

```mlir
scf.while (%arg0 = %init) : (i32) -> i64 {
  %val = .., %arg0 : i64
  %cond = arith.cmpi .., %arg0 : i32
  scf.condition(%cond) %val : i64
} do {
^bb0(%arg1: i64):
  %next = .., %arg1 : i32
  scf.yield %next : i32
}
```

Can be transformed into:

``` mlir
%pre_val = .., %init : i64
%pre_cond = arith.cmpi .., %init : i32
scf.if %pre_cond -> i64 {
  %res = scf.while (%arg1 = %va0) : (i64) -> i64 {
    // Original after block
    %next = .., %arg1 : i32
    // Original before block
    %val = .., %next : i64
    %cond = arith.cmpi .., %next : i32
    scf.condition(%cond) %val : i64
  } do {
  ^bb0(%arg2: i64):
    %scf.yield %arg2 : i32
  }
  scf.yield %res : i64
} else {
  scf.yield %pre_val : i64
}
```

The test pass for `wrapWhileLoopInZeroTripCheck` has been modified to use the new pattern when `forceCreateCheck=false`.

### CODE REVIEW

File: mlir/test/lib/Dialect/SCF/TestSCFWrapInZeroTripCheck.cpp
Comment (victor-eds) [inline mlir/test/lib/Dialect/SCF/TestSCFWrapInZeroTripCheck.cpp:1]: This test pass is no longer needed as we expose the function as a pass already.

File: mlir/include/mlir/Dialect/SCF/Transforms/Patterns.h [hunk 1]
```diff
@@ -85,6 +86,10 @@ void populateSCFForLoopCanonicalizationPatterns(RewritePatternSet &patterns);
 ///  * `after` block containing arith.addi
 void populateUpliftWhileToForPatterns(RewritePatternSet &patterns);
 
+/// Populate patterns to rotate `scf.while` ops, constructing `do-while` loops
+/// from `while` loops.
+void populateSCFRotateWhileLoopPatterns(
+    RewritePatternSet &patterns, const SCFRotateWhileLoopPassOptions &options);
```
Comment (victor-eds) [inline mlir/include/mlir/Dialect/SCF/Transforms/Patterns.h:92]: This currently has a limitation: we cannot apply the pattern recursively on newly created ops in a pass when `forceCreateCheck`. I left it as is as I am not aware of what's the "canonical" approach here. I'm using `GreedyRewriteStrictness::ExistingOps` in the pass, but passes calling this may not wanna do that.

We could either document that here or add an anonymous attribute to output `scf.while` operations to avoid infinite recursion.

Comment (victor-eds) [reply mlir/include/mlir/Dialect/SCF/Transforms/Patterns.h]: I think annotating the operation might be a better option as it gives callers more flexibility. In my use-case, I don't care much as I won't be using `forceCreateCheck`, but others might be interested.

File: mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  void runOnOperation() final {
+    Operation *parentOp = getOperation();
+    MLIRContext *context = &getContext();
+    RewritePatternSet patterns(context);
+    SCFRotateWhileLoopPassOptions options{forceCreateCheck};
+    scf::populateSCFRotateWhileLoopPatterns(patterns, options);
+    // Avoid applying the pattern to a loop more than once.
+    GreedyRewriteConfig config;
+    config.strictMode = GreedyRewriteStrictness::ExistingOps;
+    [[maybe_unused]] LogicalResult success =
```
Comment (Hardcode84) [inline mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp:68]: nit:` (void)applyPatternsAndFoldGreedily`

File: mlir/include/mlir/Dialect/SCF/Transforms/Patterns.h (hunk 1 above)
Comment (Hardcode84) [reply mlir/include/mlir/Dialect/SCF/Transforms/Patterns.h]: Do you want to use pass itself or you plan to use `populateSCFRotateWhileLoopPatterns` to combine with other patterns?

If you only need pass, using direct loop like in test pass instead of greedy rewriter may be a better solution.

Comment (victor-eds) [reply mlir/include/mlir/Dialect/SCF/Transforms/Patterns.h]: I'm planning on using the patterns, that's why I went with this. We could also hide the problem by not allowing options in the patterns. The pass can remain as is. WDYT?

File: mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                         ArrayRef<StringRef> generatedNames = {})
+      : OpRewritePattern<scf::WhileOp>(context, benefit, generatedNames),
+        forceCreateCheck(rotateLoop) {}
+
+  LogicalResult matchAndRewrite(scf::WhileOp whileOp,
+                                PatternRewriter &rewriter) const final {
+    FailureOr<scf::WhileOp> result =
+        scf::wrapWhileLoopInZeroTripCheck(whileOp, rewriter, forceCreateCheck);
+    if (failed(result) || *result == whileOp) {
+      LLVM_DEBUG(whileOp->emitRemark("Failed to rotate loop"));
```
Comment (joker-eph) [inline mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp:46]: Since this is a pattern, it can be tried over and over and over leading to repeating messages.

Also what is the behavior if the loop is already wrapped?

File: mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    Operation *parentOp = getOperation();
+    MLIRContext *context = &getContext();
+    RewritePatternSet patterns(context);
+    SCFRotateWhileLoopPassOptions options{forceCreateCheck};
+    scf::populateSCFRotateWhileLoopPatterns(patterns, options);
+    // Avoid applying the pattern to a loop more than once.
+    GreedyRewriteConfig config;
+    config.strictMode = GreedyRewriteStrictness::ExistingOps;
+    [[maybe_unused]] LogicalResult success =
+        applyPatternsAndFoldGreedily(parentOp, std::move(patterns), config);
```
Comment (joker-eph) [inline mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp:69]: Why do we need the fixed-point property of the greedy rewriter here?

The TestSCFWrapInZeroTripCheck.cpp didn't.

File: mlir/include/mlir/Dialect/SCF/Transforms/Patterns.h (hunk 1 above)
Comment (Hardcode84) [reply mlir/include/mlir/Dialect/SCF/Transforms/Patterns.h]: Yeah, I feel not exposing this option on patterns is better solution, but you will need to keep old test pass to be able to test it.

File: mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp (hunk 4 above)
Comment (victor-eds) [reply mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp:69]: I'll reply to both your comments here:

If the option forcing guard insertion even when the loop is already in a do-while shape is passed, we'd run into an infinite loop unless we have this fixpoint.

The test wasn't using patterns, but a walker on the `scf.while` operations. I wanted to expose this as a pattern. As I'm gonna get rid of the option in the populate patterns API, this won't be an issue there.

Comment (joker-eph) [reply mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp]: My comments stand though, I still don't understand why the pass is using the pattern and greedy rewriter in the first place?

for my other comment: `wrapWhileLoopInZeroTripCheck` does not document when it is returning a failure, which means I can't really understand the pattern behavior (when is it returning success/failure), can you document this?

Comment (victor-eds) [reply mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp]: I'm just walking on operations for the pass and not accepting the option in the patterns to fix potential infinite recursion issues now. Also documented when the function fails (it does not fail as of now) and added some more comments all around.

File: mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+namespace mlir {
+#define GEN_PASS_DEF_SCFROTATEWHILELOOPPASS
+#include "mlir/Dialect/SCF/Transforms/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+
+namespace {
+struct RotateWhileLoopPattern : OpRewritePattern<scf::WhileOp> {
+  using OpRewritePattern<scf::WhileOp>::OpRewritePattern;
```
Comment (Hardcode84) [inline mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp:28]: This pattern is essentially untested now. Better solution, I think, will be to keep old test pass as is and use it to test `wrapWhileLoopInZeroTripCheck` directly, and also have this new pass (without `forceCreateCheck` option) to test pattern approach specifically. Also, as you are only interested in exposing patterns, this new pass can be a test pass as well.

File: mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp (hunk 4 above)
Comment (Hardcode84) [reply mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp]: The goal, if I understood correctly, is to expose this transformation as a pattern (probably to be composable with other patterns), and this pass only needed to test it. We can keep it as test pass I think (see my other comment).

**Comment by victor-eds** [comment]
I've addressed your comments, keeping the test file to check the case in which check insertion isn't forced.

**Review by Hardcode84 — APPROVED** [review]
LGTM, but please wait for @joker-eph

File: mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  LogicalResult matchAndRewrite(scf::WhileOp whileOp,
+                                PatternRewriter &rewriter) const final {
+    // Setting this option would lead to infinite recursion on a greedy driver
+    // as 'do-while' loops wouldn't be skipped.
+    constexpr bool forceCreateCheck = false;
+    FailureOr<scf::WhileOp> result =
+        scf::wrapWhileLoopInZeroTripCheck(whileOp, rewriter, forceCreateCheck);
+    // scf::wrapWhileLoopInZeroTripCheck hasn't yet implemented a failure
+    // mechanism. 'do-while' loops are simply returned unmodified. In order to
+    // stop recursion, we check input and output operations differ.
```
Comment (joker-eph) [inline mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp:40]: Thanks for the comment here, this is what I was wondering about before!

Can you just add the appropriate documentation here as well: https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/SCF/Transforms/Transforms.h#L195-L199

Basically saying that if the original operation is returned we guarantee the IR hasn't been modified?

File: mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp (hunk 4 above)
Comment (joker-eph) [reply mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp]: > We can keep it as test pass I think (see my other comment).

Yes it should stay a test pass if the motivation is just to expose the transformation as a pattern (please adjust the PR title and description as well)

**Review by joker-eph — CHANGES_REQUESTED** [review]
From the comment thread, it's not clear why the pass is moved out of the test directory?

**Comment by victor-eds** [comment]
> From the comment thread, it's not clear why the pass is moved out of the test directory?

I decided it may be useful to create a pass to expose the pattern. Testing that pass covers `wrapWhileLoopInZeroTripCheck` with `forceCreateCheck=false`, as that's the only behavior the pass and the pattern expose. The test pass is kept, but just to check  `forceCreateCheck=true`.

**Comment by joker-eph** [comment]
> > From the comment thread, it's not clear why the pass is moved out of the test directory?
> 
> I decided it may be useful to create a pass to expose the pattern. Testing that pass covers `wrapWhileLoopInZeroTripCheck` with `forceCreateCheck=false`, as that's the only behavior the pass and the pattern expose. The test pass is kept, but just to check `forceCreateCheck=true`.

Sorry I don't understand your answer, it does not seem you're addressing the question actually: why did you need to "create a pass" when there was already one that could just be updated? Why should this pass be anywhere else than inside the Transform directory?

**Comment by victor-eds** [comment]
> Sorry I don't understand your answer, it does not seem you're addressing the question actually: why did you need to "create a pass" when there was already one that could just be updated? 

The *original* pass was a test pass under `test/lib/Dialect/SCF`. This pass was created to test the function I'm using for the implementation of the new pass. As my new pass already covers the case for `forceCreateCheck=false`, I limited the scope of this already existing *test* pass.

> Why should this pass be anywhere else than inside the Transform directory?

My new pass isn't a test pass and it's located under `Dialect/SCF/Transforms`.

**Comment by joker-eph** [comment]
> > Why should this pass be anywhere else than inside the Transform directory?
> 
> My new pass isn't a test pass and it's located under `Dialect/SCF/Transforms`.

Sorry I meant to ask "Why should this pass be anywhere else than inside the Test directory?"

Basically: unless there is a strong rationale and motivation that I miss here, I'd like to see the existing test pass used and extended instead of creating a new one here.

**Comment by victor-eds** [comment]
@joker-eph dropped introduced pass and modified test one to use patterns if `forceCheck=false`, i.e., if the pattern can be used to build the pass.

**Review by joker-eph — APPROVED** [review]

File: mlir/test/Dialect/SCF/wrap-while-loop-in-zero-trip-check.mlir [hunk 7]
```diff
@@ -20,7 +20,7 @@ func.func @wrap_while_loop_in_zero_trip_check(%bound : i32) -> i32 {
 // CHECK-SAME:      %[[BOUND:.*]]: i32) -> i32 {
 // CHECK-DAG:     %[[C0:.*]] = arith.constant 0 : i32
 // CHECK-DAG:     %[[C5:.*]] = arith.constant 5 : i32
-// CHECK-DAG:     %[[PRE_COND:.*]] = arith.cmpi slt, %[[C0]], %[[BOUND]] : i32
+// CHECK-DAG:     %[[PRE_COND:.*]] = arith.cmpi sgt, %[[BOUND]], %[[C0]] : i32
```
Comment (joker-eph) [inline mlir/test/Dialect/SCF/wrap-while-loop-in-zero-trip-check.mlir:23]: Actually I don't quite get this change: the condition `(c0 < bound)` isn't the same as `(bound < c0)`?

**Review by joker-eph — CHANGES_REQUESTED** [review]

File: mlir/test/Dialect/SCF/wrap-while-loop-in-zero-trip-check.mlir (hunk 7 above)
Comment (victor-eds) [reply mlir/test/Dialect/SCF/wrap-while-loop-in-zero-trip-check.mlir:23]: `arith.cmpi` folding moves constants to the RHS :)

Comment (joker-eph) [reply mlir/test/Dialect/SCF/wrap-while-loop-in-zero-trip-check.mlir:23]: Do you know why is your change triggering this?

But also I'm still unsure why this folding is correct?

Comment (victor-eds) [reply mlir/test/Dialect/SCF/wrap-while-loop-in-zero-trip-check.mlir:23]: `c0 < bound iff bound > c0` works. If we were using `sge` instead, we'd end up with `bound >= c0`. This introduces the case in which `bound = c0`, which is not in the original `c0 < bound` precondition. The folding pattern is correct.

Comment (joker-eph) [reply mlir/test/Dialect/SCF/wrap-while-loop-in-zero-trip-check.mlir:23]: Or right somehow I had in mind to provide the reverse condition, not the same condition with the reversed arguments... Alright time to go to bed!

**Review by joker-eph — APPROVED** [review]

File: mlir/test/Dialect/SCF/wrap-while-loop-in-zero-trip-check.mlir (hunk 7 above)
Comment (victor-eds) [reply mlir/test/Dialect/SCF/wrap-while-loop-in-zero-trip-check.mlir:23]: It happened to me for a couple of minutes too! This can be tricky sometimes. Thanks!
